src/lib/snapper/snapshot.rb

```diff
@@ -0,0 +1,199 @@
+          new_by_type(attrs)
+        end
+      end
+
+      def find(number, configs = nil, communication = default_communication)
+        all(configs, communication).find { |s| s.number == number }
+      end
+
+      def new_by_type(attrs)
+        types[attrs[:type]].new(attrs)
+      end
+
+      def get_modified_files(from, to, communication = default_communication)
+        communication.get_modified_files(from, to)
+      end
+
+      def list_configs(communication = default_communication)
+        communication.list_configs
+      end
+
+      def get_config(config, communication = default_communication)
+        communication.get_config(config)
+      end
+    end
+
+    # Generic constructor
+    def initialize(attrs = {})
+      raise "No instantiable class" if self.class == Yast::Snapshot
+      @number = attrs[:number]
+      @timestamp = attrs[:timestamp]
+      @user = attrs[:uid]
+      # CleanUp algorithm
+      @cleanup = attrs[:cleanup] || "timeline"
+      @description = attrs[:description] || ""
+      @user_data = attrs[:user_data] || {}
+      @communication =
+        attrs[:communication] || Snapshot.default_communication(self)
+      @config = attrs[:config] || Snapshot.list_configs
+    end
+
+    def save
+      if number.nil?
+        create
+      else
+        update
+      end
+    end
+
+    def name
+      "Snapshot"
+    end
+
+    def valid?
+      true
+    end
+
+    def pre?
+      false
+    end
+
+    def post?
+      false
+    end
+
+    def single?
+      false
+    end
+
+    def user_data_to_s
+      user_data.map { |k, v| "#{k}=#{v}" }.join(", ")
+    end
+
+    def date
+      timestamp.strftime("%F %T")
+    end
+
+    def update(attrs = {})
+      log.info "Updating #{inspect} with #{attrs.inspect}"
+      attrs.each { |k, v| send("#{k}=", v) }
+
+      communication.update
+
+      true
+    end
+
+    def mount_point
+      communication.mount_point
+    end
+
+    def create(communication = self.communication)
+      num = communication.create
+      num ? Snapshot.find(num, config).number : nil
+    end
+
+    def delete(communication = self.communication)
+      communication.delete
+    end
+  end
+
+  class SingleSnapshot < Snapshot
+    def name
+      "Single"
+    end
+
+    def single?
+      true
+    end
+  end
+
+  class PreSnapshot < Snapshot
+    def name
+      "Pre"
+    end
+
+    def pre?
+      true
+    end
+
+    def post(configs = nil, custom_communication = communication)
+      Snapshot.all(configs, custom_communication).find { |s| s.pre_number == number }
+    end
+  end
+
+  class PostSnapshot < Snapshot
+    def name
+      "Post"
+    end
+
+    def initialize(attrs = {})
+      @pre_number = attrs[:pre_number]
+
+      super(attrs)
+    end
+
+    def post?
+      true
+    end
+
+    def valid?
+      if !pre
+        log.info "There is no PreSnapshot for PostSnapshot #{inspect}"
+        return false
+      end
+      true
```

Comment:
I expect here something like super...if we extend validation in base class, it should also check post snapshots.